### DISCUSSION TITLE
Is refreshing session in the API example correct?

### DISCUSSION THREAD

**Discussion by derun-info** [bluesky-social/atproto#1618] [How to do it?]
Hi everyone,

The example of the API looks like this.

```typescript
await agent.login({ identifier: 'alice@mail.com', password: 'hunter2' })
await agent.resumeSession(savedSessionData);
```

I struggle to understand it. Doesn't it mean that the login will be attempted anyway (and crash from rate limit), regardless if I have saved session data?

Should it be resume session first if you have it? If it throws an exception - attempt login.

In my app I used to just login, but since today I hit rate limit whenever I restart the app during quick development, so I am looking on how to resume the session correctly.

Thanks!

**Comment by tweinmann** [comment]
Yes, calling agent.login() always creates a new session, i.e. you should only call agent.resumeSession() if you have a valid session.

**Reply by derun-info** [reply]
So pseudocode should be 

```
if (savedSession) {
  try {
     resume(savedSession)
  ) catch (e) {
     login() 
  }
} else {
  login()
}
```

Is that correct?

**Reply by tweinmann** [reply]
yes, this is correct imo. it does not count to the session rate limit for me when done this way.

**Reply by derun-info** [reply]
Thanks! Is there a file in source to check these limits? What is the current login rate limit? This morning if the first time I am hitting it. Is it per IP?

**Reply by tweinmann** [reply]
afaik it's per user and the policy is '30;w=300', i.e. 30 times in 300 minutes

**Reply by MedeHurGang** [reply]
actually it should be 30/300 seconds and 300/24 hours. but implementation seem to mix this up. actually its 30/24 hours for sessionCreate.

**Reply by derun-info** [reply]
Does it average out smoothly, like 1 per hour roughly, or if you depleted 30, you need to wait for 24 hours?

**Reply by MedeHurGang** [reply]
you have to wait until the reset timestamp in the header after using 30 sessionCreate requests.

**Reply by derun-info** [reply]
Right, but headers are managed by the API client. I wonder if the API could be improved to incorporate it into the error message.

**Reply by MedeHurGang** [reply]
as far as i know this headers come from the server not the local api.
if you assign agent.login() to a variable eg.
`var log = await agent.login({...});`
the json in log will have the headers, but this does not work for agent.resumeSession().
i havent found a way to get the headers when using the resumeSession method.

**Reply by derun-info** [reply]
This implies the function finishes, not throws the exception  with rate limit exceeded

**Reply by MedeHurGang** [reply]
maybe you need to upgrade the npm package. i have another bot with older npm version which doesn't return headers.
for me agent.login() returns something like this:
`XRPCResponse {
  data: {
    did: '*******',
    handle: '*******',
    email: '*********,
    accessJwt: '******',
    refreshJwt: '********'
  },
  headers: {
    'access-control-allow-origin': '*',
    connection: 'keep-alive',
    'content-length': '650',
    'content-type': 'application/json; charset=utf-8',
    date: 'Sun, 17 Sep 2023 12:16:44 GMT',
    etag: 'W/"28a-dVXfVeIebmu5jhj3yAMab5m8TNU"',
    'ratelimit-limit': '30',
    'ratelimit-policy': '30;w=300',
    'ratelimit-remaining': '28',
    'ratelimit-reset': '1695039404',
    vary: 'Accept-Encoding',
    'x-powered-by': 'Express'
  },
  success: true
}`

**Reply by MedeHurGang** [reply]
> This implies the function finishes, not throws the exception with rate limit exceeded

then you might allready hit the ratelimit and have to wait until it resets or use another account for development meanwhile.

**Reply by derun-info** [reply]
I am running the latest the latest `0.6.*` and it throws an error during `await login`: `error: Uncaught (in promise) Error: Rate Limit Exceeded`. It's deno, but it should not impact the choice between throwing an exception and returning an error.

Here is a test that confirms that it must throw: https://github.com/bluesky-social/atproto/blob/17ce804ffaaafb1b1efb61174217a1da275feda8/packages/xrpc-server/tests/rate-limiter.test.ts#L197

**Comment by MedeHurGang** [comment]
yes, just call login once. than resume. but at the moment, resume counts as create as well as login. so you won't win any request by using resumeSession instead of login method. at least atm.

**Reply by derun-info** [reply]
Thanks. Ptal at the pseudocode I've posted above, is it correct?

Are you saying that even if I use the `resumeSession` correctly, I won't fix the rate exceeded problem? If yes, how is it possible to do any development, where you change a few lines and restart the app?

**Reply by MedeHurGang** [reply]
for the code in my experience you just need to login once and save the session eg. to file or db. after that session get renewed with resumeSession if timed out. i tried with 12 hour old session, which got updated.

yes, thats correct. even resume lowers the ratelimit by one what i found out so far. it makes development on livesystem very time consuming atm. i filed a bugreport (https://github.com/bluesky-social/atproto/issues/1616) on this but did not get any response yet. but i think they gonna fix this soon.

**Reply by MedeHurGang** [reply]
maybe upvoting or confirming the bugreport i filed might speed this up. but i dont know, i don't even know if this is the right place to file bugs like this.

**Comment by bnewbold** [comment]
The example in the package README was not intended to show that one of the `createAccount`, `resumeSession`, or `login` methods needs to be called, not a combination of them.

We will update the example to be clearer.

**Reply by tweinmann** [reply]
I have created PR with adapted sample -> https://github.com/bluesky-social/atproto/issues/1619
